api/python/t4/snapshots.py

```diff
@@ -131,10 +139,350 @@ def download_file_from_snapshot(src, dst, snapshothash):
+
+class Package(object):
+    """ In-memory representation of a package """
+
+    def __init__(self, data=None, meta=None):
+        """
+        _data is of the form {logical_key: PackageEntry}
+        physical_keys is a list of objects of the form {
+            schema_version: string
+            type: string
+            uri: string
+        }
+        hash is of the form {
+            type: string (e.g. "SHA256")
+            value: string
+        }
+        size is the length of the object in bytes
+        """
+        self._data = data or {}
+        self._meta = meta or {}
+
+    def _clone(self):
+        """
+        Returns clone of this package.
+        """
+        return Package(copy.deepcopy(self._data), copy.deepcopy(self._meta))
+
+    def __contains__(self, logical_key):
+        """
+        Checks whether the package contains a specified logical_key.
+
+        Returns:
+            True or False
+        """
+        return logical_key in self._data
+
+    @staticmethod
+    def load(path):
+        """
+        Loads a package from a path.
+
+        Args:
+            path(string): the location to load the package from
+
+        Returns:
+            a new package object
+
+        Raises:
+            file not found
+            json decode error
+            invalid package exception
+        """
+        data = {}
+        with jsonlines.open(path) as reader:
+            meta = reader.read()
+            for obj in reader:
+                lk = obj.pop('logical_key')
+                if lk in data:
+                    raise PackageException("Duplicate logical key while loading package")
+                data[lk] = PackageEntry(
+                    obj['physical_keys'],
+                    obj['size'],
+                    obj['hash'],
+                    obj['meta']
+                )
+
+        return Package(data, meta)
+
+    @staticmethod
+    def create_package(path):
+        """
+        Takes a package of a provided path.
+
+        Recursively enumerates every file in path, and returns a new
+        package that contains all those files.
+
+        Args:
+            path(string): path to package
+
+        Returns:
+            A new Package of that path
+
+        Raises:
+            when path doesn't exist
+        """
+        # TODO: anything but local paths
+        # TODO: deserialization metadata
+        data = {}
+        meta = {}
+        src_path = pathlib.Path(path)
+        files = src_path.rglob('*')
+        for f in files:
+            if not f.is_file():
+                continue
+
+            hash_obj = {
+                'type': 'SHA256',
+                'value': hash_file(f)
+            }
+            size = os.path.getsize(f)
+            physical_keys = [{
+                'type': PhysicalKeyType.LOCAL.name,
+                'path': os.path.abspath(f)
+            }]
+            entry = PackageEntry(physical_keys, size, hash_obj, {})
+            data[str(f)] = entry
```

Comment:
I think logical keys should be relative to `src_path`?

Reply:
I'm not sure what you mean here

Reply:
If I'm doing `Package.create_package('/home/dima/tmp/'), I don't want every logical key to start with `/home/dima/tmp`, right?

Also, related to this: on Windows, you'll have logical keys with backslashes...

Reply:
oh I see what you're saying